common/src/IO/Tokenizer.h

```diff
@@ -78,23 +78,21 @@ namespace TrenchBroom {
         public:
             using Token = TokenTemplate<TokenType>;
         private:
-            using StatePtr = std::shared_ptr<TokenizerState>;
-
             class SaveState {
             private:
-                StatePtr m_state;
+                TokenizerState* m_target;
                 TokenizerState m_snapshot;
             public:
-                explicit SaveState(StatePtr state) :
-                m_state(state),
-                m_snapshot(m_state->snapshot()) {}
+                explicit SaveState(TokenizerState* target) :
```

Comment:
Ok, yeah this fits our coding style.
Was reading https://herbsutter.com/2020/02/23/references-simply/ recently though, which has some arguments on why to only use references for function arguments.

Reply:
Yeah, that article is written from the perspective of a language designer. He argues that references were put into the language for very few very specific use cases, and it makes evolving the language more complicated that references are used in all manner of ways.

When I see a reference, it communicates two things to me: "Code using this reference does not expect it to be null or to change" and "Code using this reference does not take ownership of the reference object". In such a situation, I prefer using references over pointers. Pointers, to me, are for nullable and reassignable references and for memory allocations.

In hindsight, I'd change a lot of pointers in the codebase to references. But it's too late now ;-)

Reply:
I agree with your points. I find the automatic dereferencing is the really troublesome part - I don't like that `m_bar = xyz;` might actually be modifying some random object, if m_bar is declared `Something& m_bar`.

Clion mitigates this problem somewhat by rendering a `&` character in front of reference typed variables, but then only if you're inside Clion.

Reply:
I see your point. I can't recall ever having seen this cause a bug, but maybe it's because I wrote most of the code and am already aware of this potential pitfall.

Reply:
Ok, one more point in favor of pointers that I just thought of, or maybe you mentioned it and I didn't see it before: Performance wise, it's good to be aware of dereferencing, and references hide that. So you're happily using an object and thrashing your cache with ever being aware.